### PR TITLE
Support Observable

### LINKED ISSUE #84: Support Observable
Re https://github.com/mochajs/mocha/issues/1898

-

https://github.com/zenparsing/es-observable

@thejameskyle What do we need to do to support it?

I would guess calling `.forEach()` on the observable as that returns a promise.

### PR DESCRIPTION
**Pull request by BarryThePenguin**
Fixes #84

Still working through it. Happy to hear any feedback. I've just duplicated the tests for Promises. I started out writing tests using RxJs which all passed successfully. Switched to `zen-observable` half way through so some tests are currently skipped as a result.

### CODE REVIEW

File: test/test.js [hunk 1]
```diff
@@ -96,6 +97,25 @@ test('plan assertions with support for promises', function (t) {
 	});
 });
```
Comment (sindresorhus) [inline test/test.js]: Might we worth extracting all the Observable related test into a separate test file, since there are so many now.

**Comment by sindresorhus** [comment]
Generally looks good :)

Will need docs and tests are currently failing.

File: test/test.js (hunk 1 above)
Comment (BarryThePenguin) [inline test/test.js]: Ok, and the Promise related tests too? I'm thinking the end result would be:
- test.js
- test.Promise.js
- test.Observable.js

**Comment by BarryThePenguin** [comment]
Happy to write some documentation. I'll keep looking at the failing tests, I'm not sure what the problem is yet...

**Comment by BarryThePenguin** [comment]
I think 'es6-symbol' is causing the tests to fail. What is the preferred solution for Symbols in 0.10 and 0.12?

**Comment by BarryThePenguin** [comment]
Hmm... also zen-observable requires a Promise polyfill...

**Comment by jamestalmage** [comment]
We will have to run zen-observables through the Babel `runtime` plugin. As well as anything using Symbol.

File: test/test.js (hunk 1 above)
Comment (sindresorhus) [inline test/test.js]: Yeah, but that should be done in a separate PR.

- test.js
- promise.js
- observable.js

**Comment by jamestalmage** [comment]
@BarryThePenguin #200 - will fix the Node 0.10 error - it exists on master (I don't know how the latest master build passed)

**Comment by BarryThePenguin** [comment]
Great, I'll squash and rebase again once that gets in

File: test/test.js [hunk 2]
```diff
@@ -8,6 +8,18 @@ var Runner = require('../lib/runner');
...
+	Object.defineProperty(global, 'Promise', {
+		value: Promise,
+		configurable: true,
+		enumerable: false,
+		writable: true
+	});
+}
+
+var Observable = require('zen-observable');
+
```
Comment (BarryThePenguin) [inline test/test.js]: I'm not sure if this is a Great Idea™

Any opinions?...

Comment (jamestalmage) [inline test/test.js]: Nah - just set it. That doesn't do anything to prevent pollution of globals.

Comment (BarryThePenguin) [inline test/test.js]: It's in test/test.js for Observables only as we want a lightweight implementation of the observable spec to test with. `zen-observable` "[Requires ES6 Promises or a Promise polyfill](https://github.com/zenparsing/zen-observable#zen-observable)" and this looked like a simple way to do that without also having a Symbol polyfil at the same time.

Maybe @blesh or @zenparsing have a suggestion?

Comment (blesh) [inline test/test.js]: I'm unsure of what the goal is here for "Supporting Observables". What do you need them to do? Do you need to support subscription to them? Do you need to support returning them?

If it's the former, you should be able to test for `obj[Symbol.observable]` and call it (like a function) to get back an `observable` object with a `subscribe()` function on it that follows the `zen-observable` spec. For that, there's really no need to import a library.

On the other hand, if you want to emit them, you'll want to use the zen-observable polyfill or the RxJS 5 implementation as they follow the spec. And you'll need to do something like you're doing here.

**Comment by blesh** [comment]
Okay, I think I'm caught up with what the desire is here.

Observables can "end" in one of three ways: 
They terminate because: 

1. the producer errors.
2. the producer completes successfully.
3. the consumer unsubscribed.


You only really need to care about 1 and 2, because 3 would be caused by the test itself. To handle 1 and 2 for zen-observable, RxJS 5+, Kefir, Most and Bacon, you'd just need:

```js
let obj = getSomethingThatMightBeObservable();

if (obj[Symbol.observable]) {
  obj[Symbol.observable]().subscribe({
    complete() { /* done successfully */ },
    error(err) { reportError(err); /* done because error */ }
  });
}
```

But the problem there is while you've determined that it's "done", you haven't analyzed the output.

If the idea is to analyze the output, the best short-term solution is to subscribe to `RxJS 5` or `zen-observable` Observables with `forEach`, which returns a promise:

```js
return myObservable.forEach(x => doSomeAssertion(x === 'whatever'));
```

There would be no mileage from Kefir, Most or Bacon from that unless the spec changes to include that `forEach` must be implemented on returns from `[Symbol.observable]()`. cc/ @zenparsing

**Comment by blesh** [comment]
> There would be no mileage from Kefir, Most or Bacon from that unless the spec changes to include that forEach must be implemented on returns from [Symbol.observable]()

... actually in that case it wouldn't matter anyhow, because the libraries that aren't RxJS 5 or zen-observable would need to be subscribed to in a different way.

I think that using `forEach` is probably your safest bet. It returns a promise, so you wouldn't have to make a change.

**Comment by sindresorhus** [comment]
> I think that using forEach is probably your safest bet. It returns a promise, so you wouldn't have to make a change.

:+1:

**Comment by sindresorhus** [comment]
OT to this PR, but kinda weird that you have to call `Observable.forEach(function () {});` to convert it to a promise. I would have preferred something like `Promise.resolve(observable)`.

**Comment by blesh** [comment]
> I would have preferred something like Promise.resolve(observable).

Good luck with that one. It's not a bad idea, but I think it will be hard to get traction on changes to Promise to support Observable.

However, with RxJS, you can just use `Observable.prototype.toPromise()` which will convert the Observable to a Promise that resolves with the last value of the Observable when the Observable completes, or reject with the Observable's error if it errors.

**Comment by blesh** [comment]
Also, I think it would be `Promise.from(observable)` ;)

**Comment by sindresorhus** [comment]
@BarryThePenguin Can you rebase from master and fix the merge conflict? This looks good to me.

@vdemedes @jamestalmage Anything final you'd like to point out?

**Comment by vdemedes** [comment]
Everything looks good to me, very straightforward, thanks @BarryThePenguin!

**Comment by BarryThePenguin** [comment]
:+1:

**Comment by BarryThePenguin** [comment]
Feel free to cancel [build 168](https://ci.appveyor.com/project/sindresorhus/ava/build/168)

That was a mistake

**Comment by sindresorhus** [comment]
> Feel free to cancel build 168

Done.

**Comment by sindresorhus** [comment]
Landed! Thank you @BarryThePenguin :)

**Comment by BarryThePenguin** [comment]
:tada:

**Comment by sindresorhus** [comment]
I've added some docs for it https://github.com/sindresorhus/ava/commit/e57eff059f73e53f1c8db1af096dc53e2608a322, but I don't really know how Observables work, so any improvements welcome! For example, can you `await` an Observable in an async function?

// @thejameskyle

**Comment by zenparsing** [comment]
@sindresorhus You can't directly `await` an observable.  We've discussed it at length and implicit conversion to promise for use by `await` is controversial.  You can consume the sequence with `forEach` and `await` the returned promise, or we might add something like `observable.toPromise()` in the future.

**Comment by sindresorhus** [comment]
> We've discussed it at length and implicit conversion to promise for use by await is controversial.

That's too bad... I was hoping `await` could be used directly by any async primitive. Do you happen to have a link to that discussion?

In that case, `observable.toPromise()` is definitely needed. The `forEach`-to-convert-to-promise thing is super weird.
